### PR TITLE
Add JsConfigs view helper for sharing configs with Javascript code

### PR DESCRIPTION
**Pull request by ThoWagen**
I am planning to add some JS specific configuration. This PR allows similarly to the `VuFind.translate` to access configurations in JS.

### CODE REVIEW

**Review by demiankatz — COMMENTED** [review]
@ThoWagen, does it make sense to have Javascript-related configuration in a global .ini file? Would it make more sense to put this somewhere in theme.config.php in case it needs to vary by theme?

**Comment by ThoWagen** [comment]
It depends. The option I want to add will allow one to configure if the statuses in the result list will be loaded batch wise or asynchronous for every record. And also if the all statuses will be loaded at once or only if observed. Both configs are not really theme specific but we have to add them since our ILS does not support batch wise loading and the current implementation is much slower than loading each status asynchronously.

But I could also see theme specific configuration being added. What do you think about having `getJsConfig` merging configurations of the global js.ini and theme.config.php?

**Comment by demiankatz** [comment]
@ThoWagen, I'm not opposed in theory to the idea of merging configurations as you propose, the specific use case you mention feels like something that belongs in the existing `[Item_Status]` section of config.ini. Might it make more sense to provide a more targeted configuration there instead of building a generic JS-specific mechanism?

**Comment by ThoWagen** [comment]
Yes, i think you are right. We could add a js-config.phtml like js-translation.phtml and specify which parts of the config we want to have available in JS.

**Comment by demiankatz** [comment]
Makes sense to me! Feel free to do that here, or to close this and open a separate PR later; whatever is most convenient for your needs!

**Comment by ThoWagen** [comment]
@demiankatz I changed the implementation as discussed.

File: themes/bootstrap3/templates/layout/js-configs.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/layout/js-configs.phtml]: I realize this is probably here as a placeholder for when we add actual configurations... but should we perhaps include a commented-out example in the meantime, or else remove it until it is needed?

File: themes/bootstrap3/js/config.js
Comment (demiankatz) [inline themes/bootstrap3/js/config.js:13]: Might we want a mechanism for default values if keys are unset, or is the expectation that we'll deal with all that externally?

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: This might benefit from a different name, since it doesn't necessarily convert to an array -- it could return a string or null. Maybe `convertToPhpNativeType` would be better?

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: Could this be simplified to `$value !== null && $value !== []`?

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: Do we want/need to handle the case of a non-existent config file? Maybe I'm forgetting something that smooths out the process, but I'm guessing that might lead to an error here.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen, I think this is shaping up pretty nicely. See below for a few comments. I also wonder whether we should look for some simple use cases to apply immediately so we're not introducing this mechanism without an immediate obvious purpose.

File: themes/bootstrap3/templates/layout/js-configs.phtml
Comment (ThoWagen) [reply themes/bootstrap3/templates/layout/js-configs.phtml]: I added an commented-out example. But I also plan to create the followup PR where this is used pretty soon after this is merged.

File: themes/bootstrap3/js/config.js
Comment (ThoWagen) [reply themes/bootstrap3/js/config.js:13]: I would handle this externally like "VuFind.config.get('config')?.Site?.title ?? 'Default Title'".

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: I changed the name to convertToJsonSerializable since that is what that method is used for.

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: Yes, I changed it.

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: `$this->configLoader->get($configFile);` returns an empty `Config` if the file does not exist. So this would not lead to an error.

File: themes/bootstrap3/js/config.js
Comment (demiankatz) [reply themes/bootstrap3/js/config.js:13]: Makes sense, thanks!

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: Excellent name, thanks! :-)

File: module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/View/Helper/Root/JsConfigs.php]: Great! I should have double-checked, but thanks for confirming.

**Review by demiankatz — APPROVED** [review]
Thanks, @ThoWagen, this looks good to me now. Since it's a fairly significant new feature, I'll leave it open a little longer before merging just in case anyone else has additional feedback.

**Review by demiankatz — CHANGES_REQUESTED** [review]
I just had a conversation with @EreMaijala about this PR, since I wanted another opinion before merging it. He raised some valid points that we might want to pause and think about:

1.) A mechanism this generic is potentially dangerous -- a poorly-thought-out customization could inadvertently lead to disclosure of private configurations and/or significant bloat through the export of more configurations than desired. If we go forward with this, do we need some safety mechanisms to reduce these risks?

2.) Is it worth reconsidering the scope of this? Do we expect to have enough configurations globally affecting Javascript to need it? Might it be more targeted to have a search status-specific helper that exposes settings via data elements embedded in the search templates rather than doing things this way? While I do see the attraction of having Javascript configuration access that parallels PHP configuration access, it seems like a potentially heavy-weight mechanism for limited uses. Sticking with the existing config() view helper or more purpose-built solutions may be safer and more efficient.

Anyway, I'm certainly open to hearing arguments for moving this forward in more or less its present form -- but given the potential risks involved, I want to make sure we're on the right path. And apologies for not thinking of these things a little sooner; I was focused on reviewing the technical aspects of the code (which is, in and of itself, well designed) without giving enough thought to the broader implications. :-)

**Comment by ThoWagen** [comment]
You are right that this makes it to easy to accidentally include configs that should not be exposed or are unnecessary. But I think it is still valuable to have a config.js or a similar mechanism that allows to add configurations to JS globally. In the case of the configurations for checking item statuses there are 6 templates that would need the extra data elements and possibly more in custom themes.

What do you think about adding the configuration as a simple key value mapping. E.g.in js-configs.phtml

```
<?php /* Global context config only */

$this->jsConfigs()->add(
    'myConfig',
    $this->config()->getMyConfig()
);

echo $this->jsConfigs()->getJSON();
```

that can be accessed like this:

```
if (VuFing.config('myConfig') ?? true) {
...
```

Alternatively, configs can be added in other templates only when needed.

With this there is also more freedom and control over where the configs are coming from. One could also directly set a config like:

```
$this->jsConfigs()->add('myConfig2', true);
```

**Comment by demiankatz** [comment]
@ThoWagen, I think your proposed solution of setting values more directly in `js-config.ptml` makes sense to me. I don't expect this mechanism to be used so much that there's a lot of benefit to have parallelism between PHP configs and Javascript configs, so simplifying like this seems beneficial -- and as you say, this provides a useful and extensible mechanism for making specific settings globally available, since there are occasional use cases for such things.

@EreMaijala, what do you think?

**Comment by demiankatz** [comment]
I set a 10.0 milestone here when I thought this was going to be merged right away; I'm now taking it off since this may end up taking a little while to discuss, and I don't want to make the remaining 10.0 development slate too full. I'm still happy to merge this in time for 10.0 if we happen to get the work done by then; I just don't want to _require_ us to get it done that fast.

File: themes/bootstrap3/templates/layout/js-configs.phtml
Comment (EreMaijala) [inline themes/bootstrap3/templates/layout/js-configs.phtml]: I think that we still need a warning here not to expose any sensitive settings. Something like `WARNING: Any settings here are exposed to all users in plain text, so do not include any secrets here.`

Comment (EreMaijala) [reply themes/bootstrap3/templates/layout/js-configs.phtml]: Perhaps also a couple more examples of the different ways `addConfigPaths` can be called?

**Review by EreMaijala — CHANGES_REQUESTED** [review]

File: themes/bootstrap3/templates/layout/js-configs.phtml
Comment (ThoWagen) [reply themes/bootstrap3/templates/layout/js-configs.phtml]: I added some explanation and the warning. 

I also added another example. Since this got change to a simple key value store I think two examples should be sufficient.

**Comment by ThoWagen** [comment]
@demiankatz I implemented the changes we discussed.

**Review by demiankatz — APPROVED** [review]
Looks good to me now; thanks, @ThoWagen. I'll wait for @EreMaijala's final approval before merging.

**Review by EreMaijala — APPROVED** [review]

**Pull request merged**
